traffic_ops/traffic_ops_golang/trafficstats/trafficstats.go

```diff
@@ -0,0 +1,691 @@
+		off, err := strconv.ParseUint(offset, 10, 64)
+		if err != nil {
+			e = errors.New("Invalid offset!")
+			return
+		}
+		c.Offset = &off
+	}
+
+	if orderby, ok := i.Params["orderby"]; ok {
+		if c.OrderBy = orderableFromString(orderby); c.OrderBy == nil {
+			e = errors.New("Invalid orderby!")
+			return
+		}
+	}
+
+	if c.Start, e = parseTime(i.Params["startDate"]); e != nil {
+		log.Errorf("Parsing startDate: %v", e)
+		e = errors.New("Invalid startDate!")
+		return
+	}
+
+	if c.End, e = parseTime(i.Params["endDate"]); e != nil {
+		log.Errorf("Parsing endDate: %v", e)
+		e = errors.New("Invalid endDate!")
+		return
+	}
+
+	if interval, ok := i.Params["interval"]; !ok {
+		c.Interval = DEFAULT_INTERVAL
+	} else if c.Interval = tc.TrafficStatsDurationFromString(interval); c.Interval == tc.InvalidDuration {
+		log.Errorf("Error parsing 'interval' query parameter: %v", e)
+		e = errors.New("Invalid interval!")
+		return
+	}
+
+	if ex, ok := i.Params["exclude"]; ok {
+		switch ex {
+		case "series":
+			c.ExcludeSeries = true
+		case "summary":
+			c.ExcludeSummary = true
+		default:
+			e = errors.New("Invalid exclude! Must be 'series' or 'summary'")
+			return
+		}
+	}
+
+	c.MetricType = i.Params["metricType"]
+	if _, ok := metricTypes[c.MetricType]; !ok {
+		e = fmt.Errorf("Unknown metric type: %s", c.MetricType)
+		return
+	}
+
+	var ok bool
+	if c.DeliveryService, ok = i.Params["deliveryServiceName"]; !ok {
+		if c.DeliveryService, ok = i.Params["deliveryService"]; !ok {
+			e = errors.New("You must specify deliveryService or deliveryServiceName!")
+			return
+		}
+
+		if dsID, err := strconv.ParseUint(c.DeliveryService, 10, 64); err != nil {
+			// sql.ErrNoRows does not *necessarily* mean the DS doesn't exist - an XMLID can simply
+			// be numeric, and so it was wrong to treat it as an ID in the first place.
+			xmlid := c.DeliveryService
+			if c.DeliveryService, err = getXMLIDFromID(dsID, i.Tx.Tx); err != nil && err != sql.ErrNoRows {
+				log.Errorf("Converting DSID to XMLID: %v", err)
+				e = errors.New("Internal Server Error")
+				code = http.StatusInternalServerError
+				return
+			} else if err == sql.ErrNoRows {
+				c.DeliveryService = xmlid
+			}
+		}
+	}
+
+	code = http.StatusOK
+	return
+}
+
+func GetDSStats(w http.ResponseWriter, r *http.Request) {
+	// Perl didn't require "interval", but it would only return summary data if it was not given
+	inf, userErr, sysErr, errCode := api.NewInfo(r, []string{"metricType", "startDate", "endDate"}, nil)
+	tx := inf.Tx.Tx
+	if userErr != nil || sysErr != nil {
+		api.HandleErr(w, r, tx, errCode, userErr, sysErr)
+		return
+	}
+	defer inf.Close()
+
+	var c config
+	if c, userErr, errCode = configFromRequest(r, inf); userErr != nil {
+		sysErr = fmt.Errorf("Unable to process deliveryservice_stats request: %v", userErr)
+		api.HandleErr(w, r, tx, errCode, userErr, sysErr)
+		return
+	}
+
+	client, err := inf.CreateInfluxClient()
+	if err != nil {
+		if err == sql.ErrNoRows {
```

Comment:
tbh from a caller point of view it is really weird that sql.ErrNoRows represents no servers from a function called `CreateInfluxClient` and also should this be a 500 ? 503 to me means the api is no available not another service the service relies on

Reply:
The HTTP spec seems to indicate that 503 errors are for scheduled maintenance and unplanned overloads, so I agree it's not ideal. I just hesitate to send back a 500 because that tends to make people think "there's something wrong with this endpoint, a 500 shouldn't happen". But I don't have a better idea, I suppose.